File: benchmarks/reverse-complement.rb
alu = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA".bytes

total = 3000000

seq = Array.new(total, 0)
alu_len = alu.length
for i in 0...total do
    seq[i] = alu[i % alu_len]
end

comp = Array.new(256, 0)
for i in 0...256 do
    comp[i] = i
end
comp['A'.ord] = 'T'.ord
comp['a'.ord] = 'T'.ord
comp['C'.ord] = 'G'.ord
comp['c'.ord] = 'G'.ord
comp['G'.ord] = 'C'.ord
comp['g'.ord] = 'C'.ord
comp['T'.ord] = 'A'.ord
comp['t'.ord] = 'A'.ord

t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

for iter in 0...15 do
    i = 0
    j = total - 1
    while i <= j do
        ci = seq[i]
        cj = seq[j]
        seq[i] = comp[cj]
        seq[j] = comp[ci]
        i += 1
        j -= 1
    end
end

t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)

puts ">ONE Homo sapiens alu"
k = 0
while k < total do
    end_idx = k + 60
    end_idx = total if end_idx > total
    puts seq[k...end_idx].map(&:chr).join
    k += 60
end

puts "bench_ns=#{t1 - t0}"

File: benchmarks/reverse-complement.js
const alu = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

const total = 3000000;

const seq = Buffer.allocUnsafe(total);
const aluBuf = Buffer.from(alu);
for (let i = 0; i < total; i++) {
    seq[i] = aluBuf[i % aluBuf.length];
}

const complement = Buffer.allocUnsafe(256);
for (let i = 0; i < 256; i++) complement[i] = i;
complement['A'.charCodeAt(0)] = 'T'.charCodeAt(0);
complement['a'.charCodeAt(0)] = 'T'.charCodeAt(0);
complement['C'.charCodeAt(0)] = 'G'.charCodeAt(0);
complement['c'.charCodeAt(0)] = 'G'.charCodeAt(0);
complement['G'.charCodeAt(0)] = 'C'.charCodeAt(0);
complement['g'.charCodeAt(0)] = 'C'.charCodeAt(0);
complement['T'.charCodeAt(0)] = 'A'.charCodeAt(0);
complement['t'.charCodeAt(0)] = 'A'.charCodeAt(0);

const t0 = process.hrtime.bigint();

for (let iter = 0; iter < 15; iter++) {
    let i = 0;
    let j = total - 1;
    while (i <= j) {
        const ci = seq[i];
        const cj = seq[j];
        seq[i] = complement[cj];
        seq[j] = complement[ci];
        i++;
        j--;
    }
}

const t1 = process.hrtime.bigint();

console.log(">ONE Homo sapiens alu");
for (let k = 0; k < total; k += 60) {
    let end = k + 60;
    if (end > total) end = total;
    console.log(seq.subarray(k, end).toString());
}
console.log(`bench_ns=${Number(t1 - t0)}`);
